### PR TITLE
feat(components): Add ComboBox

### PR DESCRIPTION
**Pull request by ArthurKnaus**
https://github.com/getsentry/sentry/assets/7033940/9311b7c6-6029-4c27-8e97-94d18b054fe3

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -2157,6 +2207,17 @@
     "@swc/helpers" "^0.5.0"
     clsx "^2.0.0"
 
+"@react-aria/focus@^3.16.2":
```
Comment (scttcper) [inline yarn.lock:2210]: seeing lots of duplicates

File: static/app/components/stories/sizingWindow.tsx [hunk 2]
```diff
@@ -26,4 +27,5 @@ const SizingWindow = styled(NegativeSpaceContainer)<Props>`
       `}
 `;
 
+export type Props = ComponentProps<typeof SizingWindow>;
```
Comment (scttcper) [inline static/app/components/stories/sizingWindow.tsx:30]: avoid exporting anything called props. is this used?

File: static/app/components/comboBox/index.tsx [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+  });
+
+  // Sync input value with value prop
+  const previousValue = useRef(value);
+  if (previousValue.current !== value) {
+    const selectedLabel = options
+      .flatMap(item => ('options' in item ? item.options : [item]))
+      .find(option => option.value === value)?.label;
+    if (selectedLabel) {
+      setInputValue(selectedLabel);
```
Comment (scttcper) [inline static/app/components/comboBox/index.tsx:215]: setState inside render function

Comment (ArthurKnaus) [reply static/app/components/comboBox/index.tsx:215]: Afaik this is the recommended approach for updating state from props:
https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes

Did we have problems with that?

File: static/app/components/stories/sizingWindow.tsx (hunk 2 above)
Comment (ArthurKnaus) [reply static/app/components/stories/sizingWindow.tsx:30]: Good catch! Just moved it and did not pay attention. It is used only in https://github.com/getsentry/sentry/blob/master/static/app/components/stories/matrix.tsx
Will rename 👍

File: yarn.lock (hunk 1 above)
Comment (ArthurKnaus) [reply yarn.lock:2210]: true! any recommended tool we normally use for de-duping?

File: static/app/components/comboBox/index.tsx (hunk 3 above)
Comment (scttcper) [reply static/app/components/comboBox/index.tsx:215]: when is it used? The parent component is trying to filter things for you and then you can also edit the filter?

Comment (ArthurKnaus) [reply static/app/components/comboBox/index.tsx:215]: No, the parent component controls the selected value and I need to sync the inputs text with it, when it changes.

File: yarn.lock (hunk 1 above)
Comment (ArthurKnaus) [reply yarn.lock]: Used `yarn-deduplicate`

**Comment by scttcper** [comment]
i think the input maybe isn't disabled on close, have trouble reopening before clicking outside

https://github.com/getsentry/sentry/assets/1400464/12d6e14a-7c0c-4f31-b719-385d46b4853e

**Comment by scttcper** [comment]
I think on close it should stop being an input?
image


It seems to grow forever
image

**Review by scttcper — APPROVED** [review]
let @vuluongj20 have a chance to review as well

File: static/app/components/comboBox/comboBox.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+          onChange={({value: newValue}) => setValue(newValue)}
+          aria-label="ComboBox"
+          menuTrigger="focus"
+          placeholder="Select an Option"
+          isLoading
+          options={options}
+        />
+      </SizingWindow>
+    );
+  });
```
Comment (vuluongj20) [inline static/app/components/comboBox/comboBox.stories.tsx:83]: There's an extra divider in this story:
image

File: static/app/components/comboBox/index.tsx [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+function ComboBox<Value extends string>({
+  size = 'md',
+  className,
+  placeholder,
+  disabled,
+  isLoading,
+  sizeLimitMessage,
+  menuTrigger = 'focus',
+  ...props
+}: ComboBoxProps<Value>) {
```
Comment (vuluongj20) [inline static/app/components/comboBox/index.tsx:64]: I'd also add a `leadingItems` and `trailingItems` prop, similar to what we have for `MenuListItem`. That way we can add things like leading search icon, trailing chevron, etc.
Screenshot 2024-04-09 at 3 39 33 PM

Comment (vuluongj20) [reply static/app/components/comboBox/index.tsx:64]: If you wanna make it really nice we can also use a composable structure like with `InputGroup`:
https://github.com/getsentry/sentry/blob/7e22a005f94b6316140495d2225ba99161440941/static/app/components/inputGroup.spec.tsx#L30-L35

File: static/app/components/comboBox/index.tsx [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+            )}
+            <ListBox
+              {...listBoxProps}
+              ref={listBoxRef}
+              listState={state}
+              keyDownHandler={() => true}
+              size={size}
+              sizeLimitMessage={sizeLimitMessage}
+            />
+            <EmptyMessage>No items found</EmptyMessage>
```
Comment (vuluongj20) [inline static/app/components/comboBox/index.tsx:174]: Uneven padding here:
image

File: static/app/components/comboBox/index.tsx (hunk 5 above)
Comment (ArthurKnaus) [reply static/app/components/comboBox/index.tsx:63]: good idea! I created [an issue](https://github.com/getsentry/sentry/issues/68592) for it. Will do it in a follow up!

**Pull request merged**